File: utttcpp/src/utttcpp/selfplay/monteCarloTreeSearch.cpp
#include "utttcpp/selfplay/monteCarloTreeSearch.hpp"

#include <algorithm>
#include <cmath>
#include <deque>
#include <stdexcept>

#include "utttcpp/game/constants.hpp"
#include "utttcpp/helpers/prngMersenneTwister.hpp"

namespace utttcpp {

const std::string serializeEvaluatedState(EvaluatedState *evaluatedState) {
    std::string serializedEvaluatedState = "evaluatedState{";
    for (int i = 0; i < STATE_SIZE; i++) {
        serializedEvaluatedState += std::to_string(static_cast<int>(evaluatedState->state[i]));
    }
    serializedEvaluatedState += " ";
    serializedEvaluatedState += std::to_string(evaluatedState->numVisits);
    serializedEvaluatedState += " ";
    serializedEvaluatedState += std::to_string(evaluatedState->numWins);
    serializedEvaluatedState += " ";
    serializedEvaluatedState += std::to_string(evaluatedState->numDraws);
    serializedEvaluatedState += " ";
    serializedEvaluatedState += std::to_string(evaluatedState->numLosses);
    serializedEvaluatedState += "}";
    return serializedEvaluatedState;
}

const std::string serializeEvaluatedActions(EvaluatedAction **evaluatedActions,
                                            int numEvaluatedActions) {
    std::string serializedEvaluatedActions = "evaluatedActions{";
    for (int i = 0; i < numEvaluatedActions; i++) {
        if (i > 0) {
            serializedEvaluatedActions += ",";
        }
        serializedEvaluatedActions += std::to_string(static_cast<int>(evaluatedActions[i]->symbol));
        serializedEvaluatedActions += " ";
        serializedEvaluatedActions += std::to_string(evaluatedActions[i]->index);
        serializedEvaluatedActions += " ";
        serializedEvaluatedActions += std::to_string(evaluatedActions[i]->numVisits);
        serializedEvaluatedActions += " ";
        serializedEvaluatedActions += std::to_string(evaluatedActions[i]->numWins);
        serializedEvaluatedActions += " ";
        serializedEvaluatedActions += std::to_string(evaluatedActions[i]->numDraws);
        serializedEvaluatedActions += " ";
        serializedEvaluatedActions += std::to_string(evaluatedActions[i]->numLosses);
    }
    serializedEvaluatedActions += "}";
    return serializedEvaluatedActions;
}

Node::Node(const UltimateTicTacToe *uttt, const Action action) : uttt(uttt), action(action) {
    childNodes = nullptr;
    childNodesLength = 0;
    numVisits = 0;
    numWinsX = 0;
    numWinsO = 0;
    numDraws = 0;
}

Node::~Node() {
    if (childNodes != nullptr) {
        for (int i = 0; i < childNodesLength; i++) {
            delete childNodes[i];
        }
        delete[] childNodes;
    }
    delete uttt;
}

void Node::expand() {
    if (!isLeaf()) {
        return;
    }
    if (isTerminal()) {
        return;
    }
    std::vector<Action> legalActions = uttt->getLegalActions();
    childNodesLength = legalActions.size();
    childNodes = new Node *[childNodesLength];
    for (int i = 0; i < childNodesLength; i++) {
        Action childAction = legalActions[i];
        UltimateTicTacToe *childUttt = uttt->clone();
        childUttt->execute(childAction);
        Node *childNode = new Node(childUttt, childAction);
        childNodes[i] = childNode;
    }
}

EvaluatedState *Node::getEvaluatedState() const {
    int numWins, numLosses;
    if (uttt->isNextSymbolX()) {
        numWins = numWinsX;
        numLosses = numWinsO;
    } else {
        numWins = numWinsO;
        numLosses = numWinsX;
    }
    EvaluatedState *evaluatedState = new EvaluatedState{
        new unsigned char[STATE_SIZE],
        numVisits,
        numWins,
        numDraws,
        numLosses,
    };
    const unsigned char *utttState = uttt->getState();
    for (int i = 0; i < STATE_SIZE; i++) {
        evaluatedState->state[i] = utttState[i];
    }
    return evaluatedState;
}

EvaluatedAction **Node::getEvaluatedActions() const {
    if (isLeaf()) {
        throw std::runtime_error("cannot get evaluated actions from a leaf node");
    }
    if (isTerminal()) {
        throw std::runtime_error("cannot get evaluated actions from a terminal node");
    }
    EvaluatedAction **evaluatedActions = new EvaluatedAction *[childNodesLength];
    int numWins, numLosses;
    for (int i = 0; i < childNodesLength; i++) {
        if (childNodes[i]->action.isSymbolX()) {
            numWins = childNodes[i]->numWinsX;
            numLosses = childNodes[i]->numWinsO;
        } else {
            numWins = childNodes[i]->numWinsO;
            numLosses = childNodes[i]->numWinsX;
        }
        EvaluatedAction *evaluatedAction = new EvaluatedAction{
            childNodes[i]->action.symbol,
            childNodes[i]->action.index,
            childNodes[i]->numVisits,
            numWins,
            childNodes[i]->numDraws,
            numLosses,
        };
        evaluatedActions[i] = evaluatedAction;
    }
    return evaluatedActions;
}

std::string Node::toString() const {
    std::string nodeString = "Node(\n";
    nodeString += "  action: " + action.toString() + "\n";
    nodeString += "  uttt: ";
    std::string utttString = uttt->toString();
    for (std::string::size_type i = 0, j = 0; j < utttString.length(); j++) {
        if (utttString[j] == '\n') {
            nodeString += utttString.substr(i, 1 + j - i);
            nodeString += "  ";
            i = j + 1;
        } else if (j + 1 == utttString.length()) {
            nodeString += utttString.substr(i, 1 + j - i);
        }
    }
    nodeString += "\n";
    nodeString += "  num_children: " + std::to_string(childNodesLength) + "\n";
    nodeString += "  num_visits: " + std::to_string(numVisits) + "\n";
    nodeString += "  num_X_wins: " + std::to_string(numWinsX) + "\n";
    nodeString += "  num_O_wins: " + std::to_string(numWinsO) + "\n";
    nodeString += "  num_draws: " + std::to_string(numDraws) + "\n";
    nodeString += ")";
    return nodeString;
}

Tree::Tree(Node *node) : root(node) {}

Tree::~Tree() { delete root; }

void Tree::synchronize(const UltimateTicTacToe *uttt) {
    int nextRootIndex = -1;
    for (int i = 0; i < root->childNodesLength; i++) {
        Node *childNode = root->childNodes[i];
        if (uttt->isEqualTo(childNode->uttt)) {
            nextRootIndex = i;
            break;
        }
    }
    Node *prevRoot = root;
    if (nextRootIndex != -1) {
        root = root->childNodes[nextRootIndex];
        for (int i = 0; i < prevRoot->childNodesLength; i++) {
            if (i != nextRootIndex) {
                delete prevRoot->childNodes[i];
            }
            prevRoot->childNodes[i] = nullptr;
        }
        delete[] prevRoot->childNodes;
        prevRoot->childNodes = nullptr;
    } else {
        root = new Node(uttt->clone(), Action());
    }
    delete prevRoot;
}

std::string Tree::toString() const {
    std::string treeString = "Tree(\n";
    treeString += "  root: ";
    std::string rootString = root->toString();
    for (std::string::size_type i = 0, j = 0; j < rootString.length(); j++) {
        if (rootString[j] == '\n') {
            treeString += rootString.substr(i, 1 + j - i);
            treeString += "  ";
            i = j + 1;
        } else if (j + 1 == rootString.length()) {
            treeString += rootString.substr(i, 1 + j - i);
        }
    }
    treeString += "\n";
    treeString += "  size: " + std::to_string(getSize()) + "\n";
    treeString += "  height: " + std::to_string(getHeight()) + "\n";
    treeString += ")";
    return treeString;
}

int Tree::bfsCountNodes(Node *node) const {
    int cnt = 0;
    std::deque<Node *> nodes = {node};
    while (!nodes.empty()) {
        Node *node = nodes.front();
        nodes.pop_front();
        cnt += 1;
        for (int i = 0; i < node->childNodesLength; i++) {
            nodes.push_back(node->childNodes[i]);
        }
    }
    return cnt;
}

int Tree::dfsMaxDepth(Node *node, int depth) const {
    if (node->childNodesLength == 0) {
        return depth;
    }
    int maxDepth = depth;
    for (int i = 0; i < node->childNodesLength; i++) {
        maxDepth = std::max(maxDepth, dfsMaxDepth(node->childNodes[i], depth + 1));
    }
    return maxDepth;
}

MonteCarloTreeSearch::MonteCarloTreeSearch(UltimateTicTacToe *uttt,
                                           int numSimulations,
                                           double explorationStrength)
    : tree(new Tree(new Node(uttt, Action()))),
      numSimulations(numSimulations),
      explorationStrength(explorationStrength) {}

MonteCarloTreeSearch::~MonteCarloTreeSearch() { delete tree; }

void MonteCarloTreeSearch::run() {
    const int numRunSimulations = numSimulations - tree->root->numVisits;
    for (int i = 0; i < numRunSimulations; i++) {
        simulate();
    }
}

Action MonteCarloTreeSearch::selectAction(EvaluatedAction **evaluatedActions,
                                          int numEvaluatedActions,
                                          const std::string selectionMethod) const {
    if (!(numEvaluatedActions > 0)) {
        throw std::runtime_error("numEvaluatedActions should be > 0");
    }
    EvaluatedAction *selectedEvaluatedAction = nullptr;
    if (selectionMethod == "argmax") {
        int *bestIndices = new int[numEvaluatedActions];
        int bestIndicesLength = 0;
        int bestScore = 0;
        for (int i = 0; i < numEvaluatedActions; i++) {
            if (evaluatedActions[i]->numVisits > bestScore) {
                bestIndicesLength = 0;
                bestIndices[bestIndicesLength++] = i;
                bestScore = evaluatedActions[i]->numVisits;
            } else if (evaluatedActions[i]->numVisits == bestScore) {
                bestIndices[bestIndicesLength++] = i;
            }
        }
        if (bestIndicesLength == 1) {
            selectedEvaluatedAction = evaluatedActions[bestIndices[0]];
        } else {
            selectedEvaluatedAction =
                evaluatedActions[bestIndices[randInt(0, bestIndicesLength - 1)]];
        }
        delete[] bestIndices;
    } else if (selectionMethod == "sample") {
        int totalNumVisits = 0;
        for (int i = 0; i < numEvaluatedActions; i++) {
            totalNumVisits += evaluatedActions[i]->numVisits;
        }
        if (!(totalNumVisits > 0)) {
            throw std::runtime_error("totalNumVisits should be > 0");
        }
        const int randomThreshold = randInt(1, totalNumVisits);
        int cumulativeNumVisits = 0;
        for (int i = 0; i < numEvaluatedActions; i++) {
            cumulativeNumVisits += evaluatedActions[i]->numVisits;
            if (randomThreshold <= cumulativeNumVisits) {
                selectedEvaluatedAction = evaluatedActions[i];
                break;
            }
        }
        if (selectedEvaluatedAction == nullptr) {
            throw std::runtime_error("selectedEvaluatedAction should be defined here");
        }
    } else if (selectionMethod == "random") {
        const int randomIndex = randInt(0, numEvaluatedActions - 1);
        selectedEvaluatedAction = evaluatedActions[randomIndex];
    } else {
        throw std::runtime_error("unknown selectionMethod");
    }
    return Action(selectedEvaluatedAction->symbol, selectedEvaluatedAction->index);
}

std::string MonteCarloTreeSearch::toString() const {
    std::string mctsString = "MonteCarloTreeSearch(\n";
    mctsString += "  tree: ";
    std::string treeString = tree->toString();
    for (std::string::size_type i = 0, j = 0; j < treeString.length(); j++) {
        if (treeString[j] == '\n') {
            mctsString += treeString.substr(i, 1 + j - i);
            mctsString += "  ";
            i = j + 1;
        } else if (j + 1 == treeString.length()) {
            mctsString += treeString.substr(i, 1 + j - i);
        }
    }
    mctsString += "\n";
    mctsString += "  num_simulations: " + std::to_string(numSimulations) + "\n";
    mctsString += "  exploration_strength: " + std::to_string(explorationStrength) + "\n";
    mctsString += ")";
    return mctsString;
}

void MonteCarloTreeSearch::simulate() {
    selectLeafNode();
    Node *leafNode = selectedPath[selectedPathLength - 1];
    // leafNode->expand(); (leaf node is not expanded here to improve memory efficiency)
    playoutBackprop(leafNode);
}

void MonteCarloTreeSearch::selectLeafNode() {
    Node *node = tree->root;
    int topScoreIndicesLength, topChildNodeIndex;
    double score, topScore;
    selectedPathLength = 0;
    // while (!node->isLeaf()) { (original version)
    while (node->numVisits != 0 && !node->isTerminal()) {  // accommodate for unexpanded leaf node
        if (node->numVisits == 1) {  // detect unexpanded leaf node
            node->expand();  // expand only when node's children are needed
        }
        selectedPath[selectedPathLength++] = node;
        topScoreIndicesLength = 0;
        topScore = -1e9;
        for (int i = 0; i < node->childNodesLength; i++) {
            score = UCT(node->childNodes[i], node->numVisits);
            if (score > topScore) {
                topScoreIndicesLength = 0;
                topScoreIndices[topScoreIndicesLength++] = i;
                topScore = score;
            } else if (score == topScore) {
                topScoreIndices[topScoreIndicesLength++] = i;
            }
        }
        if (topScoreIndicesLength == 1) {
            topChildNodeIndex = topScoreIndices[0];
        } else if (topScoreIndicesLength > 1) {
            topChildNodeIndex = topScoreIndices[randInt(0, topScoreIndicesLength - 1)];
        } else {
            throw std::runtime_error("topScoreIndicesLength is not determined");
        }
        node = node->childNodes[topChildNodeIndex];
    }
    selectedPath[selectedPathLength++] = node;
}

double MonteCarloTreeSearch::UCT(const Node *node, const int parentNumVisits) const {
    if (node->numVisits == 0) {
        return 1e9;
    }
    double exploitationScore = valueFunction(node);
    double explorationScore = explorationStrength *
        std::sqrt(std::log(static_cast<double>(parentNumVisits)) / node->numVisits);
    return exploitationScore + explorationScore;
}

double MonteCarloTreeSearch::valueFunction(const Node *node) const {
    if (node->action.isSymbolX()) {
        return static_cast<double>(node->numWinsX - node->numWinsO) /
               static_cast<double>(node->numVisits);
    } else if (node->action.isSymbolO()) {
        return static_cast<double>(node->numWinsO - node->numWinsX) /
               static_cast<double>(node->numVisits);
    }
    throw std::runtime_error("unknown action symbol");
}

void MonteCarloTreeSearch::playoutBackprop(const Node *node) {
    UltimateTicTacToe *uttt = node->uttt->clone();
    while (!uttt->isTerminated()) {
        unsigned char nextSymbol = uttt->getNextSymbol();
        int numLegalIndexes = uttt->findLegalIndexes(playoutLegalIndexes);
        int selectedIndex = playoutLegalIndexes[randInt(0, numLegalIndexes - 1)];
        uttt->execute(Action(nextSymbol, selectedIndex));
    }
    if (uttt->isResultX()) {
        for (int i = 0; i < selectedPathLength; i++) {
            selectedPath[i]->numVisits += 1;
            selectedPath[i]->numWinsX += 1;
        }
    } else if (uttt->isResultO()) {
        for (int i = 0; i < selectedPathLength; i++) {
            selectedPath[i]->numVisits += 1;
            selectedPath[i]->numWinsO += 1;
        }
    } else if (uttt->isResultDraw()) {
        for (int i = 0; i < selectedPathLength; i++) {
            selectedPath[i]->numVisits += 1;
            selectedPath[i]->numDraws += 1;
        }
    } else {
        throw std::runtime_error("unknown uttt result");
    }
    delete uttt;
}

}  // namespace utttcpp

File: utttpy/selfplay/monte_carlo_tree_search.py
from __future__ import annotations

import math
import random
from collections import deque
from typing import List, Optional, Tuple

from tqdm import tqdm

from utttpy.game.action import Action
from utttpy.game.ultimate_tic_tac_toe import UltimateTicTacToe


class MonteCarloTreeSearch:

    def __init__(
        self,
        uttt: UltimateTicTacToe,
        num_simulations: int,
        exploration_strength: float,
    ):
        self.tree = Tree(root=Node(uttt=uttt))
        self.num_simulations = num_simulations
        self.exploration_strength = exploration_strength

    def run(self, progress_bar: bool = False) -> None:
        num_run_simulations = self.num_simulations - self.tree.root.num_visits
        for i in tqdm(range(num_run_simulations), disable=not progress_bar):
            simulate(
                node=self.tree.root,
                exploration_strength=self.exploration_strength,
            )

    def get_evaluated_state(self) -> dict:
        return self.tree.root.get_evaluated_state()

    def get_evaluated_actions(self) -> List[dict]:
        return self.tree.root.get_evaluated_actions()

    def select_action(
        self, evaluated_actions: List[dict], selection_method: str
    ) -> Action:
        if selection_method == "argmax":
            max_num_visits = max(
                evaluated_action["num_visits"] for evaluated_action in evaluated_actions
            )
            top_evaluated_actions = [
                evaluated_action
                for evaluated_action in evaluated_actions
                if evaluated_action["num_visits"] >= max_num_visits
            ]
            selected_evaluated_action = random.choice(top_evaluated_actions)
        elif selection_method == "sample":
            num_visits_list = [
                evaluated_action["num_visits"] for evaluated_action in evaluated_actions
            ]
            total_num_visits = sum(num_visits_list)
            weights = [num_visits / total_num_visits for num_visits in num_visits_list]
            selected_evaluated_action = random.choices(evaluated_actions, weights=weights, k=1)[0]
        elif selection_method == "random":
            selected_evaluated_action = random.choice(evaluated_actions)
        else:
            raise ValueError(f"unknown selection_method={repr(selection_method)}")
        return Action(
            symbol=selected_evaluated_action["symbol"],
            index=selected_evaluated_action["index"],
        )

    def synchronize(self, uttt: UltimateTicTacToe) -> None:
        self.tree.synchronize(uttt=uttt)

    def __str__(self):
        output = (
            '{cls}(\n'
            '  tree: {tree}\n'
            '  num_simulations: {num_simulations}\n'
            '  exploration_strength: {exploration_strength}\n)'
        )
        output = output.format(
            cls=self.__class__.__name__,
            tree=str(self.tree).replace('\n', '\n  '),
            num_simulations=self.num_simulations,
            exploration_strength=self.exploration_strength,
        )
        return output


class Tree:

    def __init__(self, root: Node):
        self.root = root

    @property
    def size(self) -> int:
        return bfs_count_nodes(node=self.root)

    @property
    def height(self) -> int:
        return dfs_max_depth(node=self.root, depth=0)

    def synchronize(self, uttt: UltimateTicTacToe) -> None:
        for child_node in self.root.child_nodes:
            if uttt.is_equal_to(child_node.uttt):
                self.root = child_node
                return
        self.root = Node(uttt=uttt.clone())

    def __str__(self):
        output = (
            '{cls}(\n'
            '  root: {root}\n'
            '  size: {size}\n'
            '  height: {height}\n)'
        )
        output = output.format(
            cls=self.__class__.__name__,
            root=str(self.root).replace('\n', '\n  '),
            size=self.size,
            height=self.height,
        )
        return output


class Node:

    def __init__(self, uttt: UltimateTicTacToe, action: Optional[Action] = None):
        self.uttt = uttt
        self.action = action
        self.child_nodes = []
        self.num_visits = 0
        self.num_X_wins = 0
        self.num_O_wins = 0
        self.num_draws = 0

    def is_leaf(self) -> bool:
        return len(self.child_nodes) == 0

    def is_terminal(self) -> bool:
        return self.uttt.is_terminated()

    def expand(self) -> None:
        if not self.is_leaf():
            return
        if self.is_terminal():
            return
        legal_actions = self.uttt.get_legal_actions()
        if len(legal_actions) == 0:
            raise MonteCarloTreeSearchError("expanding node with no legal actions")
        for legal_action in legal_actions:
            uttt = self.uttt.clone()
            uttt.execute(action=legal_action, verify=False)
            child_node = Node(uttt=uttt, action=legal_action)
            self.child_nodes.append(child_node)

    def get_evaluated_state(self) -> dict:
        if self.uttt.is_next_symbol_X():
            num_wins = self.num_X_wins
            num_losses = self.num_O_wins
        else:
            num_wins = self.num_O_wins
            num_losses = self.num_X_wins
        return {
            "state": self.uttt.state.copy(),
            "num_visits": self.num_visits,
            "num_wins": num_wins,
            "num_draws": self.num_draws,
            "num_losses": num_losses,
        }

    def get_evaluated_actions(self) -> List[dict]:
        if self.is_leaf():
            raise MonteCarloTreeSearchError("node is a leaf")
        if self.num_visits == 0:
            raise MonteCarloTreeSearchError("node was not visited")
        evaluated_actions = []
        for child_node in self.child_nodes:
            if child_node.action.is_symbol_X():
                num_wins = child_node.num_X_wins
                num_losses = child_node.num_O_wins
            else:
                num_wins = child_node.num_O_wins
                num_losses = child_node.num_X_wins
            evaluated_action = {
                "symbol": child_node.action.symbol,
                "index": child_node.action.index,
                "num_visits": child_node.num_visits,
                "num_wins": num_wins,
                "num_draws": child_node.num_draws,
                "num_losses": num_losses,
            }
            evaluated_actions.append(evaluated_action)
        return evaluated_actions

    def __str__(self):
        output = (
            '{cls}(\n'
            '  action: {action}\n'
            '  uttt: {uttt}\n'
            '  num_children: {num_children}\n'
            '  num_visits: {num_visits}\n'
            '  num_X_wins: {num_X_wins}\n'
            '  num_O_wins: {num_O_wins}\n'
            '  num_draws: {num_draws}\n)'
        )
        output = output.format(
            cls=self.__class__.__name__,
            action=self.action,
            uttt=str(self.uttt).replace('\n', '\n  '),
            num_children=len(self.child_nodes),
            num_visits=self.num_visits,
            num_X_wins=self.num_X_wins,
            num_O_wins=self.num_O_wins,
            num_draws=self.num_draws,
        )
        return output


def simulate(node: Node, exploration_strength: float) -> None:
    selected_path = select_leaf_node(node=node, exploration_strength=exploration_strength)
    if len(selected_path) == 0:
        raise MonteCarloTreeSearchError("selected path is empty")
    leaf_node = selected_path[-1]
    leaf_node.expand()
    stats = playout(node=leaf_node)
    backprop(selected_path=selected_path, stats=stats)


def select_leaf_node(node: Node, exploration_strength: float) -> List[Node]:
    selected_path = []
    while not node.is_leaf():
        selected_path.append(node)
        scores = [
            UCT(
                node=child_node,
                parent_num_visits=node.num_visits,
                exploration_strength=exploration_strength,
            )
            for child_node in node.child_nodes
        ]
        top_score = max(scores)
        top_score_indices = [i for i, score in enumerate(scores) if score >= top_score]
        top_child_node_index = random.choice(top_score_indices)
        node = node.child_nodes[top_child_node_index]
    selected_path.append(node)
    return selected_path


def UCT(node: Node, parent_num_visits: int, exploration_strength: float) -> float:
    if node.num_visits == 0:
        return float("inf")
    exploitation_score = value_function(node)
    exploration_score = exploration_strength * math.sqrt(
        math.log(parent_num_visits) / node.num_visits
    )
    UCT_value = exploitation_score + exploration_score
    return UCT_value


def value_function(node: Node) -> float:
    if node.action.is_symbol_X():
        num_wins = node.num_X_wins
        num_losses = node.num_O_wins
    elif node.action.is_symbol_O():
        num_wins = node.num_O_wins
        num_losses = node.num_X_wins
    return (num_wins - num_losses) / node.num_visits


def playout(node: Node) -> Tuple[int, int, int]:
    num_X_wins = 0
    num_O_wins = 0
    num_draws = 0
    uttt = node.uttt.clone()
    while not uttt.is_terminated():
        actions = uttt.get_legal_actions()
        action = random.choice(actions)
        uttt.execute(action, verify=False)
    if uttt.is_result_X():
        num_X_wins += 1
    elif uttt.is_result_O():
        num_O_wins += 1
    elif uttt.is_result_draw():
        num_draws += 1
    return num_X_wins, num_O_wins, num_draws


def backprop(selected_path: List[Node], stats: Tuple[int, int, int]) -> None:
    num_X_wins, num_O_wins, num_draws = stats
    for node in selected_path:
        node.num_visits += 1
        node.num_X_wins += num_X_wins
        node.num_O_wins += num_O_wins
        node.num_draws += num_draws


def bfs_count_nodes(node: Node) -> int:
    cnt = 0
    nodes = deque([node])
    while len(nodes) > 0:
        node = nodes.popleft()
        cnt += 1
        nodes.extend(node.child_nodes)
    return cnt


def dfs_max_depth(node: Node, depth: int) -> int:
    if len(node.child_nodes) == 0:
        return depth
    max_depth = depth
    for child_node in node.child_nodes:
        max_depth = max(max_depth, dfs_max_depth(node=child_node, depth=depth + 1))
    return max_depth


def serialize_evaluated_state(evaluated_state: dict) -> str:
    state = "".join(map(str, evaluated_state["state"]))
    num_visits = str(evaluated_state["num_visits"])
    num_wins = str(evaluated_state["num_wins"])
    num_draws = str(evaluated_state["num_draws"])
    num_losses = str(evaluated_state["num_losses"])
    evaluated_state_str = f"evaluatedState{{{state} {num_visits} {num_wins} {num_draws} {num_losses}}}"
    return evaluated_state_str


def serialize_evaluated_actions(evaluated_actions: List[dict]) -> str:
    evaluated_actions_str = "evaluatedActions{"
    for i, evaluated_action in enumerate(evaluated_actions):
        if i > 0:
            evaluated_actions_str += ","
        symbol = str(evaluated_action["symbol"])
        index = str(evaluated_action["index"])
        num_visits = str(evaluated_action["num_visits"])
        num_wins = str(evaluated_action["num_wins"])
        num_draws = str(evaluated_action["num_draws"])
        num_losses = str(evaluated_action["num_losses"])
        evaluated_action_str = f"{symbol} {index} {num_visits} {num_wins} {num_draws} {num_losses}"
        evaluated_actions_str += evaluated_action_str
    evaluated_actions_str += "}"
    return evaluated_actions_str


class MonteCarloTreeSearchError(Exception):
    pass
